### PR TITLE
python311Packages.vispy: unbreak

### PR DESCRIPTION
**Pull request by pbsds**
## Description of changes

fixes https://hydra.nixos.org/build/265119077

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/python-modules/vispy/default.nix [hunk 1]
```diff
@@ -31,6 +31,13 @@ buildPythonPackage rec {
     hash = "sha256-77u4R6kIuvfnFpq5vylhOKOTZPNn5ssKjsA61xaZ0x0=";
   };
 
+  postPatch = ''
+    # https://numpy.org/devdocs/dev/depending_on_numpy.html#numpy-2-0-specific-advice
+    # upstream enforce builds with numpy 2+, which is backward compat with 1.xx
+    substituteInPlace pyproject.toml \
+      --replace-warn "numpy>=2.0.0rc2" "numpy"
```
Comment (kirillrdy) [inline pkgs/development/python-modules/vispy/default.nix:38]: any reason not to use `--replace-fail` ?

Comment (pbsds) [reply pkgs/development/python-modules/vispy/default.nix]: gut feeling, changed it to `fail`

File: pkgs/development/python-modules/vispy/default.nix [hunk 2]
```diff
@@ -31,6 +31,13 @@ buildPythonPackage rec {
     hash = "sha256-77u4R6kIuvfnFpq5vylhOKOTZPNn5ssKjsA61xaZ0x0=";
   };
 
+  postPatch = ''
+    # https://numpy.org/devdocs/dev/depending_on_numpy.html#numpy-2-0-specific-advice
+    # upstream enforce builds with numpy 2+, which is backward compat with 1.xx
+    substituteInPlace pyproject.toml \
+      --replace-fail "numpy>=2.0.0rc2" "numpy"
```
Comment (kirillrdy) [inline pkgs/development/python-modules/vispy/default.nix:38]: should we also change `numpy` to `numpy_2`  ( in probagatedBuildInputs )  ? or is it going to cause more issues ?

Comment (pbsds) [reply pkgs/development/python-modules/vispy/default.nix:38]: I imagine that will make it difficult to combine into a python env with other packages that rely on `numpy`.

**Review by kirillrdy — APPROVED** [review]
Result of `nixpkgs-review pr 325366` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages marked as broken and skipped:</summary>
  <ul>
    <li>napari</li>
    <li>napari.dist</li>
    <li>python312Packages.napari</li>
    <li>python312Packages.napari.dist</li>
  </ul>
</details>
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li>python311Packages.napari</li>
    <li>python311Packages.napari-svg</li>
    <li>python311Packages.napari-svg.dist</li>
    <li>python311Packages.napari.dist</li>
    <li>python311Packages.vispy</li>
    <li>python311Packages.vispy.dist</li>
    <li>python312Packages.napari-svg</li>
    <li>python312Packages.napari-svg.dist</li>
    <li>python312Packages.vispy</li>
    <li>python312Packages.vispy.dist</li>
  </ul>
</details>

**Pull request merged**